x-pack/legacy/plugins/apm/server/routes/create_api.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { merge } from 'lodash';
+import Boom from 'boom';
+import { InternalCoreSetup } from 'src/core/server';
+import { Request } from 'hapi';
+import * as t from 'io-ts';
+import { PathReporter } from 'io-ts/lib/PathReporter';
+import { ServerAPI, RouteFactoryFunction, RouteParams } from './typings';
+import { removeUndefinedProps } from '../../public/context/UrlParamsContext/helpers';
+
+type GenericRouteFactoryFunction = RouteFactoryFunction<
+  any,
+  any,
+  { [key in keyof RouteParams]?: RouteParams[key] },
+  any
+>;
+
+export function createApi() {
+  const factoryFns: GenericRouteFactoryFunction[] = [];
+  const router = {
+    add(fn: GenericRouteFactoryFunction) {
+      factoryFns.push(fn);
+      return router;
+    },
+    init: (core: InternalCoreSetup) => {
+      const { server } = core.http;
+      factoryFns.forEach(fn => {
+        const { params = {}, ...route } = fn(core);
+        server.route(
+          merge(
+            {
+              options: {
+                tags: ['access:apm']
+              }
+            },
+            route,
+            {
+              handler: async (request: Request) => {
+                const paramMap = {
+                  path: request.params,
+                  body: request.payload,
+                  query: request.query
+                };
+
+                const parsedParams = (Object.keys(params) as Array<
+                  keyof typeof params
+                >).reduce(
+                  (acc, key) => {
+                    let codec = params[key];
+                    if (!codec) return acc;
+
+                    if ('_tag' in codec) {
+                      codec = t.exact(codec);
+                    }
```

Comment:
I'm fairly new to io-ts. What is `_tag` and `t.exact`?

Reply:
t.exact() converts a `io-ts` type into an exact type, meaning no other properties can be present on the value other than those than defined in the type. By default, it's inexact, so `{a: 'a', b: 'b' }` would pass `t.type({ a: t.string })`. `_tag` is a refinement, as you can only call `exact` on types that have props (you can't call it on t.string). I'm using a better refinement in the next iteration though.